### PR TITLE
[API] Turn on/off api endpoints with parameter

### PR DESCRIPTION
**Pull request by SirDomin**
| Q               | A
| --------------- | -----
| Branch?         | 1.8
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/test/src/Tests/DisablingApiTest.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            ['auth_bearer' => $this->JWTAdminUserToken]
+        );
+
+        $this->assertResponseIsSuccessful();
+    }
+
+    /**
+     * @test
+     */
+    public function it_returns_route_not_found_with_api_disabled(): void
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/test/src/Tests/DisablingApiTest.php:45]:
```suggestion
    public function it_returns_route_not_found_if_api_disabled(): void
```
???

File: src/Sylius/Bundle/ApiBundle/Resources/config/services/resolver.xml [hunk 2]
```diff
@@ -25,6 +25,7 @@
             <argument type="service" id="api_platform.cache.route_name_resolver" />
             <argument type="service" id="api_platform.route_name_resolver.cached.inner" />
             <argument type="service" id="Sylius\Bundle\ApiBundle\Provider\PathPrefixProviderInterface" />
+            <argument>%sylius_api.enabled%</argument>
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/services/resolver.xml:28]: should we mention this change in the upgrade file?

File: src/Sylius/Bundle/ApiBundle/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -27,6 +27,12 @@ public function getConfigTreeBuilder(): TreeBuilder
         /** @var ArrayNodeDefinition $rootNode */
         $rootNode = $treeBuilder->getRootNode();
 
+        $rootNode
+            ->children()
+                ->booleanNode('enabled')->end()
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/DependencyInjection/Configuration.php:32]:
```suggestion
                ->booleanNode('enabled')
                    ->defaultFalse()
                ->end()
```
?

Comment (SirDomin) [reply src/Sylius/Bundle/ApiBundle/DependencyInjection/Configuration.php:32]: indeed, but i left it as defaultTrue

File: docs/api/unified_api/introduction.rst [hunk 4]
```diff
@@ -5,6 +5,7 @@ Introduction
 
     The new, unified Sylius API is still under development, that's why the whole ``ApiBundle`` is tagged with ``@experimental``.
     This means that all code from ``ApiBundle`` is excluded from :doc:`Backward Compatibility Promise </book/organization/backward-compatibility-promise>`.
+    You can disable entire API by changing the flag ``sylius_api.enabled`` to ``false`` in ``src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml``.
```
Comment (GSadee) [inline docs/api/unified_api/introduction.rst:8]: I guess that we should have API disabled by default:
```suggestion
    You can enable entire API by changing the flag ``sylius_api.enabled`` to ``true`` in ``app/config/packages/_sylius.yaml``.

    .. code-block:: yaml

        # app/config/packages/_sylius.yaml
        sylius_api:
            enabled: true
```

File: src/Sylius/Bundle/ApiBundle/EventSubscriber/KernelRequestEventSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        return [
+            KernelEvents::REQUEST => ['validateApi', EventPriorities::PRE_VALIDATE],
+        ];
+    }
+
+    public function validateApi(RequestEvent $event): void
+    {
+        $pathInfo = $event->getRequest()->getPathInfo();
+
+        if($this->apiEnabled === false && strpos($pathInfo, $this->apiRoute) !== false) {
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/EventSubscriber/KernelRequestEventSubscriber.php:48]:
```suggestion
        if ($this->apiEnabled === false && strpos($pathInfo, $this->apiRoute) !== false) {
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml [hunk 6]
```diff
@@ -13,6 +13,9 @@ parameters:
     sylius.security.new_api_user_account_route: "%sylius.security.new_api_shop_route%/account"
     sylius.security.new_api_user_account_regex: "^%sylius.security.new_api_user_account_route%"
 
+sylius_api:
+    enabled: true
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/app/config.yaml:17]:
```suggestion
    enabled: false
```
?

File: docs/api/unified_api/introduction.rst (hunk 4 above)
Comment (GSadee) [reply docs/api/unified_api/introduction.rst:8]: And if the API will be disabled by default, there should be a note in the UPGRADE file

File: src/Sylius/Bundle/ApiBundle/EventSubscriber/KernelRequestEventSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+namespace Sylius\Bundle\ApiBundle\EventSubscriber;
+
+use ApiPlatform\Core\EventListener\EventPriorities;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+/** @experimental  */
+class KernelRequestEventSubscriber implements EventSubscriberInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/EventSubscriber/KernelRequestEventSubscriber.php:23]:
```suggestion
final class KernelRequestEventSubscriber implements EventSubscriberInterface
```

**Review by lchrusciel — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/test/src/Tests/DisablingApiTest.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    use SetUpTestsTrait;
+
+    public function setUp(): void
+    {
+        $this->setFixturesFiles([]);
+        $this->setUpTest();
+    }
+
+    /**
+     * @test
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/test/src/Tests/DisablingApiTest.php:30]: should be in single line

**Review by AdamKasp — APPROVED** [review]

**Review by lchrusciel — COMMENTED** [review]
Let's merge it to 1.8

**Comment by AdamKasp** [comment]
Thank you, @SirDomin! :1st_place_medal:

**Comment by papatails** [comment]
And this would be me . Just saying thank you I’ve tried a lot of things and just to avail you broke them down. Now what can I do on the owner’s end. I’ve figured out when someone said just go inside and look around. I did in may 2020 life hasn’t been the same, I’m still trying and there still here but I shall conquer this with help from people like you.

**Pull request merged**